### PR TITLE
Add torchvision.MaskRCNN

### PR DESCRIPTION
**Pull request by eugene123tw**
### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have ran e2e tests and there is no issues.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Comment by harimkang** [comment]
Is this PR a releases/2.0.0 target? Or is it a develop target?

**Comment by sovrasov** [comment]
I think it's for develop

File: src/otx/recipe/instance_segmentation/maskrcnn_efficientnetb2b.yaml [hunk 1]
```diff
@@ -51,7 +51,7 @@ overrides:
                 - 1024
           - class_path: otx.core.data.transform_libs.torchvision.Pad
             init_args:
-              size_divisor: 32
+              pad_to_square: true
```
Comment (eunwoosh) [inline src/otx/recipe/instance_segmentation/maskrcnn_efficientnetb2b.yaml:54]: Don't below `side_divisor` have to be changed to `pad_to_square`? And could you tell me why all `side_divisor = 32` is changed to `pad_to_squre`?

File: src/otx/algo/instance_segmentation/maskrcnn_tv.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+            (backbone.out_channels, 7, 7),
+            [256, 256, 256, 256],
+            [1024],
+            norm_layer=nn.BatchNorm2d,
+        )
+        mask_head = MaskRCNNHeads(backbone.out_channels, [256, 256, 256, 256], 1, norm_layer=nn.BatchNorm2d)
+
+        model = TVMaskRCNN(
+            backbone,
+            num_classes=91,
```
Comment (eunwoosh) [inline src/otx/algo/instance_segmentation/maskrcnn_tv.py:90]: Is it ok to fix it?

**Review by eunwoosh — COMMENTED** [review]
Thanks for your work. I left some comments. Please take a look.

File: src/otx/algo/instance_segmentation/maskrcnn_tv.py (hunk 2 above)
Comment (eugene123tw) [reply src/otx/algo/instance_segmentation/maskrcnn_tv.py:90]: To load a pre-trained COCO model, we need to create a model with the same number of output classes as the original COCO model was trained on.

File: src/otx/recipe/instance_segmentation/maskrcnn_efficientnetb2b.yaml (hunk 1 above)
Comment (eugene123tw) [reply src/otx/recipe/instance_segmentation/maskrcnn_efficientnetb2b.yaml:54]: Updated. And the purpose is to ensure that the resizing op aligns with the OV fit-to-window resize pre-processing step.

**Review by eunwoosh — APPROVED** [review]

**Review by harimkang — APPROVED** [review]

File: src/otx/algo/instance_segmentation/maskrcnn_tv.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    ) -> None:
+        super().__init__(
+            label_info=label_info,
+            optimizer=optimizer,
+            scheduler=scheduler,
+            metric=metric,
+            torch_compile=torch_compile,
+            tile_config=tile_config,
+        )
+        self.image_size = (1, 3, 1024, 1024)
```
Comment (kprokofi) [inline src/otx/algo/instance_segmentation/maskrcnn_tv.py:58]: We need to think how we can avoid hardcoded numbers in the models. Image_size is something that user may want to change. Same for mean/std

File: src/otx/algo/instance_segmentation/maskrcnn_tv.py (hunk 2 above)
Comment (kprokofi) [reply src/otx/algo/instance_segmentation/maskrcnn_tv.py:90]: Maybe we can use strict=False? This number of classes a bit confusing.

Comment (eugene123tw) [reply src/otx/algo/instance_segmentation/maskrcnn_tv.py:90]: @kprokofi Setting `strict=False` doesn't resolve the issue.

```
model = TVMaskRCNN(
            backbone,
            num_classes=num_classes,
            rpn_anchor_generator=rpn_anchor_generator,
            rpn_head=rpn_head,
            box_head=box_head,
            mask_head=mask_head,
        )

model.load_state_dict(weights.get_state_dict(progress=True, check_hash=True))

RuntimeError: Error(s) in loading state_dict for TVMaskRCNN:
	size mismatch for roi_heads.box_predictor.cls_score.weight: copying a param with shape torch.Size([91, 1024]) from checkpoint, the shape in current model is torch.Size([2, 1024]).
	size mismatch for roi_heads.box_predictor.cls_score.bias: copying a param with shape torch.Size([91]) from checkpoint, the shape in current model is torch.Size([2]).
	size mismatch for roi_heads.box_predictor.bbox_pred.weight: copying a param with shape torch.Size([364, 1024]) from checkpoint, the shape in current model is torch.Size([8, 1024]).
	size mismatch for roi_heads.box_predictor.bbox_pred.bias: copying a param with shape torch.Size([364]) from checkpoint, the shape in current model is torch.Size([8]).
	size mismatch for roi_heads.mask_predictor.mask_fcn_logits.weight: copying a param with shape torch.Size([91, 256, 1, 1]) from checkpoint, the shape in current model is torch.Size([2, 256, 1, 1]).
	size mismatch for roi_heads.mask_predictor.mask_fcn_logits.bias: copying a param with shape torch.Size([91]) from checkpoint, the shape in current model is torch.Size([2]).
```

File: src/otx/algo/instance_segmentation/maskrcnn_tv.py (hunk 3 above)
Comment (eugene123tw) [reply src/otx/algo/instance_segmentation/maskrcnn_tv.py:58]: I agree, it's only used for export, and currently, there's no other way to get attributes from recipe configs. We could find a way to access recipe parameters, but I don't know how to do it. This should be addressed in a different PR.

**Review by kprokofi — APPROVED** [review]

File: src/otx/algo/instance_segmentation/maskrcnn_tv.py (hunk 3 above)
Comment (kprokofi) [reply src/otx/algo/instance_segmentation/maskrcnn_tv.py:58]: I will put this task to our backlog

File: src/otx/algo/instance_segmentation/maskrcnn_tv.py (hunk 2 above)
Comment (kprokofi) [reply src/otx/algo/instance_segmentation/maskrcnn_tv.py:90]: Ok, I see, let's keep it then

**Pull request merged**